src/client/opamSwitchCommand.ml

```diff
@@ -494,17 +494,57 @@ let import_t ?ask importfile t =
   end;
   t
 
-let export rt ?(full=false) filename =
-  let switch = OpamStateConfig.get_switch () in
+let freeze_opam src_dir nv = function
+  | None -> None
+  | Some opam ->
+    match OpamFile.OPAM.url opam with
+    | None -> Some opam
+    | Some url ->
+      let url_t = OpamFile.URL.url url in
+      match url_t.backend with
+      | #OpamUrl.version_control ->
+        begin match OpamProcess.Job.run (OpamGit.VCS.revision (src_dir nv)) with
+          | None ->
+            OpamConsole.warning "couldn't figure revision of %s"
+              (OpamPackage.Name.to_string nv.name);
+            None
+          | Some hash ->
+            let url_with_hash =
+              let url_t = { url_t with hash = Some hash } in
+              let checksum, mirrors =
+                OpamFile.URL.checksum url, OpamFile.URL.mirrors url
+              in
+              OpamFile.URL.create ~mirrors ~checksum url_t
+            in
+            Some (OpamFile.OPAM.with_url url_with_hash opam)
+        end
+      | `http ->
+        begin match OpamFile.URL.checksum url with
+          | [] ->
+            OpamConsole.warning "Ignoring package %s, no checksum"
+              (OpamPackage.Name.to_string nv.name);
+            None
```

Comment:
I'm a bit uneasy about this and the failure case below -- now they emit a `OpamConsole.warning`, but maybe the export should fail hard (this way, we could guarantee once frozen and full export is done, this can be imported). But: I don't know how to fail hard in this code (`export` returns a `unit`, thus there must be some exception mechanism or a direct `exit`?)